tenseal/utils/scope.h

```diff
@@ -0,0 +1,35 @@
+#ifndef TENSEAL_UTILS_SCOPE_H
+#define TENSEAL_UTILS_SCOPE_H
+
+#include <functional>
+
+namespace tenseal {
+
+class scope_guard {
+   public:
+    template <class Callable>
+    scope_guard(Callable&& undo_func) try
+        : f(std::forward<Callable>(undo_func)) {
+    } catch (...) {
+        undo_func();
+        throw;
+    }
+
+    scope_guard(scope_guard&& other) : f(std::move(other.f)) {
+        other.f = nullptr;
+    }
+
+    ~scope_guard() {
+        if (f) f();  // must not throw
+    }
+
+    void dismiss() noexcept { f = nullptr; }
+
+    scope_guard(const scope_guard&) = delete;
+    void operator=(const scope_guard&) = delete;
+
+   private:
+    std::function<void()> f;
+};
+}  // namespace tenseal
+#endif
```

Comment:
Can I know what this is doing?

Reply:
Scoped guards implement a classic C++ idiom - RAII, Resource Acquisition is Initialization. shared_ptr is another example of that.
Long story short, It will get a lambda function on the constructor, and it will call it on its destructor.
The use case is simple: I want to add some logic on a function return. For example, I want to free some resources, a create a scope guard with a lambda that frees the resources, and no matter how many returns I have in the function, the resources will get freed on exit, because the variable's destructor gets called.


My issue here was that, while writing serialization test cases, I observed that when calling make_context_public with a true value for galois or relin keys, there was a crash

This is the reason, I am to blame
https://github.com/OpenMined/TenSEAL/blame/master/tenseal/tensealcontext.cpp#L124

I was making the secret key null before creating the galois/relin keys.


The fix is to create a scoped guard that will make the secret key null on exit, no matter how many conditional returns there will be. Something like this, and I'm sure it will make the secret key null everytime
https://github.com/OpenMined/TenSEAL/blob/protobuffs/tenseal/tensealcontext.cpp#L166

I hope the explanation makes sense.

Reply:
Thanks for the explanation. It's fine to keep it in this PR, but feel free to open another one if it's convenient for you.